### ISSUE TITLE
Add support for modules

### ISSUE THREAD

**Issue by JeanJPNM** [mlogjs/mlogjs#98]
Because fery few other compilers did that. Also, this requires changes on the sourcemapping format and the name handling and resolution.

**Comment by Sharlottes** [comment]
> Because fery few other compilers did that. 

yes, so we need to make a kinds of "macro" like C's `#include` or "bundler" like webpack

**Comment by hsaturn** [comment]
I (very) often need this with mlogjs !

Why not just include a file (at the parser level) ?
`import 'module.js';`

This would just read the concatenate module.js to the rest of the file (the compiler should even not know that a module was imported).

(I'm not an expert at all in javascript)

**Comment by JeanJPNM** [comment]
> Why not just include a file (at the parser level) ?

Because mlogjs doesn't actually do the parsing by itself, it uses [babel](https://github.com/babel/babel) to parse the code and then traverses the AST (Abstract Syntax Tree) that babel generated.

The are a few things that are in the way of implementing modules:
- Implementing a module resolution algorithm (needs the path of the input file and needs to give the compiler access to the file system, instead of just giving it to the CLI)
- Implementing import views ([see this stackoverflow answer](https://stackoverflow.com/a/44157035/13745435))
- Updating the sourcemaps feature to handle multiple files
- Make all of it compatible with the web environment so that it works on the online editor 

And while module support is within my priorities, I find it more important to implement the optimizer first, as that will eliminate many bugs in the current implementation and possibly make handling modules more straightforward in some aspects.
